### ISSUE TITLE
Unresolvable ValueAfterTable errors

### ISSUE THREAD

**Issue by chrisduerr** [alexcrichton/toml-rs#258]
So I've been writing a config that should deserialize from TOML, YAML and JSON. I've started writing just the YAML bindings and everything looked good. JSON also didn't cause any issues.

However when trying to serialize my struct with `serde` and `toml-rs`, I got the `ValueAfterTable` error. I managed to resolve the initial error by reordering my struct. However when trying to write a test for the full configuration file, I wasn't able to resolve the `ValueAfterTable` errors.

This is the definition of the configuration file:
https://github.com/chrisduerr/bar-config/blob/402a4f6076ea7677cb56e311b47f5cdafed06040/src/config.rs

And here is the input I'm creating from a yaml file:
https://github.com/chrisduerr/bar-config/blob/402a4f6076ea7677cb56e311b47f5cdafed06040/tests/config.yml

I might be missing something, but I was not able to resolve the `ValueAfterTable` errors by reordering my struct.

Is this a problem that just needs a change of the structure or a custom serializer?

**Comment by ehuss** [comment]
Empty `Vec` is serialized as a value (`foo = []`) whereas if it contains structs, it is done as an array of tables.  Place the `Vec`s that contain structs at the end, and add `#[serde(skip_serializing_if = "Vec::is_empty")]` to avoid that.  (@alexcrichton is that fixable?)

You also have a few other things that need to be put at the end (like `ComponentSettings::border`).

@alexcrichton I also imagine adding the struct/key name to the `ValueAfterTable` error would make debugging this situation easier?

**Comment by alexcrichton** [comment]
Adding more info to the error message is a great idea yeah, but I'm not sure if there's much that we can do about the empty slice case. I do think we want to be sure to emit empty slices and not automatically skip all of them, and otherwise the toml serializer would need some sort of hint that it's an array of tables rather than an array of primitives, but I'm not sure if we can get that from serde :(

**Comment by chrisduerr** [comment]
@ehuss Thanks for the quick reply!

I managed to make this work by using a combination of adding the `Vec::is_empty` checks and reordering the fields in my struct.

I can imagine that this might be a massive pain especially with structs that are significantly bigger than mine. Unless you build those bottom up testing toml at every step I think that it's easy to get overwhelmed.

If this is just a toml limitation, improving the errors that are emitted would be a step in the right direction. I've used toml before and never ran into this, so maybe I just got a bit unlucky here.

Anyways, thanks for helping out! Since this wasn't really an error with toml I'm gonna go ahead and close this issue myself.
